src/cdomains/intDomain.ml

```diff
@@ -1826,9 +1890,479 @@ module Enums : S with type int_t = BigInt.t = struct
       10, QCheck.map pos (ISet.arbitrary ());
     ] (* S TODO: decide frequencies *)
 
+    let refine_with_congruence a b =
+      let contains c m x = if BI.equal m BI.zero then BI.equal c x else (BI.rem (BI.sub x c) m) == BI.zero in
+      match a with
+      | Inc e -> ( match b with
+                | None -> Inc e
+                | Some (c, m) -> Inc (ISet.filter (contains c m) e))
+      | a -> a
+  let refine_with_interval a b = a
+  let refine_with_excl_list a b = a
+  let refine_with_incl_list a b = a
+end
+
+module CongruenceFunctor(Ints_t : IntOps.IntOps): S with type int_t = Ints_t.t and type t = (Ints_t.t * Ints_t.t) option =
+struct
+  let name () = "congruences"
+  type int_t = Ints_t.t
+
+  (* represents congruence class of c mod m *)
+  type t = (Ints_t.t * Ints_t.t) option
+
+  let ( *: ) = Ints_t.mul
+  let (+:) = Ints_t.add
+  let (-:) = Ints_t.sub
+  let (%:) = Ints_t.rem
+  let (/:) = Ints_t.div
+  let (=:) = Ints_t.equal
+  let (<:) x y = Ints_t.compare x y < 0
+  let (>:) x y = Ints_t.compare x y > 0
+  let (<=:) x y = Ints_t.compare x y <= 0
+  let (>=:) x y = Ints_t.compare x y >= 0
+  (* a divides b *)
+  let ( |: ) a b = (b %: a) =: Ints_t.zero
+
+  let to_yojson t = failwith "to yojson unimplemented"
+
+  let rec gcd x y = (*TODO Double Check*)
+    if y =: Ints_t.zero then x else gcd y (x %: y)
+
+  let abs x = let r = Ints_t.neg x
+    in if x <: r then r else x
+
+  let normalize x =
+    match x with
+    | None -> None
+    | Some (c, m) -> if m =: Ints_t.zero then Some (c, m) else Some (c %: (abs m), (abs m))
+
+  let min_int ik = Ints_t.of_bigint @@ fst @@ Size.range_big_int ik
+
+  let max_int ik = Ints_t.of_bigint @@ snd @@ Size.range_big_int ik
+  let top () = Some (Ints_t.zero, Ints_t.one)
+  let top_of ik = Some (Ints_t.zero, Ints_t.one)
+  let bot () = None
+  let bot_of ik = bot () (* TODO: improve  *)
+
+  let is_top x = x = top ()
+
+  let show = function ik -> match ik with
+    | None -> "∅"
+    | Some (c, m) when (c, m) = (Ints_t.zero, Ints_t.zero) -> Ints_t.to_string c
+    | Some (c, m) ->
+      let a = if c =: Ints_t.zero then "" else Ints_t.to_string c in
+      let b = if m =: Ints_t.zero then "" else if m = Ints_t.one then "ℤ" else Ints_t.to_string m^"ℤ" in
+      let c = if a = "" || b = "" then "" else "+" in
+      a^c^b
+
+  let equal a b = match (normalize a), (normalize b) with
+    | None, None -> true
+    | Some (a, b), Some (c, d) -> a =: c && b =: d
+    | _, _ -> false
+
+  include Std (struct type nonrec t = t let name = name let top_of = top_of let bot_of = bot_of let show = show let equal = equal end)
+
+  let equal_to i = function
+    | None -> failwith "unsupported: equal_to with bottom"
+    | Some (a, b) when b =: Ints_t.zero -> if a = i then `Eq else `Neq
+    | Some (a, b) -> if i %: b = a then `Top else `Neq
+
+  let set_overflow_flag ik =
+    if Cil.isSigned ik && !GU.in_verifying_stage then
+      Goblintutil.did_overflow := true
```

Comment:
How is congruence analysis detecting overflow?

Reply:
We should not set the overflow flag here, setting it leads to reporting an overflow right away, however this is usually quite imprecise in the congruence domain.
If we are more serious about overflow checking, we would need to check for each subexpression if an active domain can exclude the overflow, and set the overflow flag only id no active domain can